### PR TITLE
Ashvin/controlpanel

### PR DESCRIPTION
**Pull request by ashvinnagarajan**
### Summary
Created control panel to interface with dyno microcontroller

#### Notes
1. Constants are defined at the top. Definitely possible to move to a separate config file if you'd prefer that. Open to ideas
2. I hate commenting out code (line 36) but i found it necessary for testing. Currently just adds one to all of the values instead of reading from the serial port.
3. Line 123 currently using two for loops to display each of the columns of labels. Im sure there is a clever way to just use one loop but i couldnt think of it. 

##### Test Plan
- [x] Linted
- [x] Tested without connection microcontroller
- [ ] Tested with microcontroller

### CODE REVIEW

**Comment by ashvinnagarajan** [comment]
image
This is what the panel currently looks like. Still need to add sections to send motor and magnet flags.

File: dyno/panda/dyno_panel.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    #Read in input from MCU sensors
+    '''for x in range (0, NUMSENSORS1): #read sensor statements
+        val = ser1.readline().decode('ascii')
+        prefix = val[0:3]
+        my_dict[prefix] = float(val[5:-2])
+    for x in range (0, NUMSENSORS2): #read JOULEMETER statements
+        val = ser2.readline().decode('ascii')
+        prefix = val[0:3]
+        my_dict[prefix] = float(val[5:-2])'''
+    for key in my_dict.keys():
```
Comment (abdullahelqaq) [inline dyno/panda/dyno_panel.py]: Hmm I see what you mean. I would still prefer the code that lands on master to be production-ready though. Maybe there's a compromise here with something like:
```suggestion
    if serial_read_on:   
        for x in range (0, NUMSENSORS1): #read sensor statements
            val = ser1.readline().decode('ascii')
            prefix = val[0:3]
            my_dict[prefix] = float(val[5:-2])
        for x in range (0, NUMSENSORS2): #read JOULEMETER statements
            val = ser2.readline().decode('ascii')
            prefix = val[0:3]
            my_dict[prefix] = float(val[5:-2])
    for key in my_dict.keys():
```
You'd have to add support for that `serial_read_on` flag though, probably by having it on by default and allowing the user to disable it when running something like:
`python dyno_panel.py --no-serial`

What are your thoughts on this?

File: dyno/panda/dyno_panel.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
+import tkinter as tk
+from tkinter import font
+from PIL import Image, ImageTk
+import serial
+import time
+
+#Define constants
+NUMSENSORS1 = 2 #RPM AND TEMPERATURE
+NUMSENSORS2 = 2 #JOULEMETER
```
Comment (abdullahelqaq) [inline dyno/panda/dyno_panel.py]: Nit: Is there any chance we can use more descriptive names here? Especially for constants that we want to be exposing to the user to configure the dyno setup. Also, is there a reason we aren't lumping both of these constants into something like a `num_runtime_inputs`? If they're handled differently, fair enough. Just wondering.

File: dyno/panda/dyno_panel.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+color_bg = "#16495C"
+color_fg = "#DA8669"
+color_label = "#B1D9DB"
+#---------------------
+
+#Placement
+LABEL_COL1_PLACE = {'relx':0.15, 'rely':0.6}
+LABEL_COL2_PLACE = {'relx':0.55, 'rely':0.6}
+LABEL_SPACE = 0.08
+#---------------------
```
Comment (abdullahelqaq) [inline dyno/panda/dyno_panel.py]: Not that it's necessary now, but eventually it might be nice to separate the frontend from the backend here into separate files. Feel free to just mark this as resolved for now, just a thing to keep in mind.

File: dyno/panda/dyno_panel.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        var_array[i].set(temp[i][0] + str(temp[i][1]) + temp[i][2])
+
+    window.after(100, update_labels)
+#---------------------
+
+#Define on_click function for button
+def handle_click(event):
+    if button["text"] == "Pause":
+        button["text"] = "Run"
+        #ser.write('P;')
```
Comment (abdullahelqaq) [inline dyno/panda/dyno_panel.py]: Yeah having commented code out in many places like this scares me a little. Ran into quite a few issues with this when writing motor control code.

File: dyno/panda/dyno_panel.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    f.write('\n')
+    f.close()
+#---------------------
+
+#Create the viewable window
+window = tk.Tk()
+window.configure(bg= color_bg)
+window.geometry("800x500")
+font_title = font.Font(family="Helvetia", size=36, weight="bold")
+font_text = font.Font(family="Helvetia", size=24)
```
Comment (abdullahelqaq) [inline dyno/panda/dyno_panel.py]: Should this be Helvetica?

File: dyno/panda/dyno_panel.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+#---------------------
+
+#Initialize variables
+my_list = []
+my_dict={
```
Comment (abdullahelqaq) [inline dyno/panda/dyno_panel.py]: Nit: Can we come up with more descriptive names here?

File: dyno/panda/dyno_panel.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+#---------------------
+
+#Create labels to display sensor readings
+for i in range(int(num_vars/2)): #First Column
+    label1 = tk.Label(window, textvariable=var_array[i], bg=color_bg, fg=color_label, font=font_text)
+    label1.place(relx = LABEL_COL1_PLACE['relx'], rely = LABEL_COL1_PLACE['rely']+i*LABEL_SPACE, anchor='w')
+
+for i in range(int(num_vars/2), num_vars): #Second Column
+    label2 = tk.Label(window, textvariable=var_array[i], bg=color_bg, fg=color_label, font=font_text)
+    label2.place(relx = LABEL_COL2_PLACE['relx'], rely = LABEL_COL2_PLACE['rely']+(i-4)*LABEL_SPACE, anchor='w')
```
Comment (abdullahelqaq) [inline dyno/panda/dyno_panel.py]:
```suggestion
offset = int(num_vars/2)
for i in range(int(num_vars/2)): #First Column
    label1 = tk.Label(window, textvariable=var_array[i], bg=color_bg, fg=color_label, font=font_text)
    label1.place(relx = LABEL_COL1_PLACE['relx'], rely = LABEL_COL1_PLACE['rely']+i*LABEL_SPACE, anchor='w')
    label2 = tk.Label(window, textvariable=var_array[i+offset], bg=color_bg, fg=color_label, font=font_text)
    label2.place(relx = LABEL_COL2_PLACE['relx'], rely = LABEL_COL2_PLACE['rely']+(i+offset-4)*LABEL_SPACE, anchor='w')
```
Please check my logic on this haha. Feel free to ignore this suggestion too if it takes away from readability. Two for loops is fine too.

**Review by abdullahelqaq — COMMENTED** [review]
Looks awesome! Just a few general comments with an eye on maintainability 💯

File: dyno/panda/dyno_panel.py (hunk 1 above)
Comment (Bruin-SuperMileage) [reply dyno/panda/dyno_panel.py]: Thats absolutely brilliant. Will definitely add

File: dyno/panda/dyno_panel.py (hunk 7 above)
Comment (Bruin-SuperMileage) [reply dyno/panda/dyno_panel.py]: Definitely works. Simplifies a bit so im down

**Comment by Bruin-SuperMileage** [comment]
I realized im logged in as BSV. All the comments above were mine. --Ashvin

File: dyno/panda/dyno_panel.py (hunk 2 above)
Comment (Bruin-SuperMileage) [reply dyno/panda/dyno_panel.py]: The panda will need to loop through each serial port separately to read in the messages. So I believe we need to specify the constants separately. Will change the names though!

File: dyno/panda/dyno_panel.py (hunk 3 above)
Comment (Bruin-SuperMileage) [reply dyno/panda/dyno_panel.py]: Yeah I was thinking about this too. Is not this entire script just frontend? I guess the functions which print and read serial can be moved to a separate script?

**Comment by ashvinnagarajan** [comment]
Made the suggested changes. Unclear variables have been renamed. Moved all the labels to one column because the controls will go on the other side.

**Comment by ashvinnagarajan** [comment]
@abdullahelqaq @anaymehta Please review on priority. Need to make other changes that build off of this stuff

File: dyno/panda/dyno_panel.py [hunk 8]
```diff
@@ -43,20 +43,20 @@ def update_labels():
         for x in range (0, LEXI_NUMSENSORS): #read Lexi sensor data
             val = lexi.readline().decode('ascii')
             prefix = val[0:3]
-            my_dict[prefix] = float(val[5:-2])
+            data_output_dict[prefix] = float(val[5:-2])
         for x in range (0, OLIVER_NUMSENSORS): #read Oliver sensor data
             val = oliver.readline().decode('ascii')
             prefix = val[0:3]
-            my_dict[prefix] = float(val[5:-2])
+            data_output_dict[prefix] = float(val[5:-2])
```
Comment (anaymehta) [inline dyno/panda/dyno_panel.py]: If speed is of the essence, I would suggest using a single string containing all the information. It's significantly faster, and not too hard to parse.

**Review by anaymehta — APPROVED** [review]
Looks good to me, new variable names are much improved.

**Review by abdullahelqaq — APPROVED** [review]
Looks gooood let's goooo

File: dyno/panda/dyno_panel.py (hunk 8 above)
Comment (ashvinnagarajan) [reply dyno/panda/dyno_panel.py]: Will add it to my backlog to take care of this. Thanks for the note
